### PR TITLE
Spell/Auras: Allow auras to be able to access gameobject caster

### PR DESCRIPTION
**Pull request by Golrag**
**Changes proposed:**

-  Allow auras to be able to access gameobject caster
-  Remove code that cleared caster guid from aura in case of gameobject
-  Add function in aura to access GameObject caster in case caster guid is of type gameobject

**Issues addressed:**

See https://github.com/TrinityCore/TrinityCore/discussions/25986#discussioncomment-1974708

**Tests performed:**
- Builds

**Known issues and TODO list:**

- [ ] Maybe we should change AuraCreateInfo Caster to a `WorldObject*` but I fear this might make the PR a whole lot bigger than initially thought

### CODE REVIEW

File: src/server/game/Spells/Auras/SpellAuras.cpp [hunk 1]
```diff
@@ -527,6 +522,14 @@ Unit* Aura::GetCaster() const
     return ObjectAccessor::GetUnit(*GetOwner(), GetCasterGUID());
 }
 
+GameObject* Aura::GetGameObjectCaster() const
```
Comment (Shauren) [inline src/server/game/Spells/Auras/SpellAuras.cpp:525]: Change this to return WorldObject and rename to GetWorldObjectCaster

Comment (Golrag) [reply src/server/game/Spells/Auras/SpellAuras.cpp]: done :)

File: src/server/game/Spells/Auras/SpellAuras.cpp [hunk 2]
```diff
@@ -249,7 +249,9 @@ void AuraApplication::BuildUpdatePacket(WorldPackets::Spells::AuraInfo& auraInfo
     // send stack amount for aura which could be stacked (never 0 - causes incorrect display) or charges
     // stack amount has priority over charges (checked on retail with spell 50262)
     auraData.Applications = aura->IsUsingStacks() ? aura->GetStackAmount() : aura->GetCharges();
-    if (!(auraData.Flags & AFLAG_NOCASTER))
+    if (aura->GetCasterGUID().IsGameObject())
```
Comment (Shauren) [inline src/server/game/Spells/Auras/SpellAuras.cpp:252]:
```suggestion
    if (!aura->GetCasterGUID().IsUnit())
```

File: src/server/game/Spells/Auras/SpellAuras.cpp [hunk 3]
```diff
@@ -392,14 +394,7 @@ Aura* Aura::Create(AuraCreateInfo& createInfo)
...
     {
-        // world gameobjects can't own auras and they send empty casterguid
-        // checked on sniffs with spell 22247
-        if (createInfo.CasterGUID.IsGameObject())
-        {
-            createInfo.Caster = nullptr;
-            createInfo.CasterGUID.Clear();
-        }
-        else
+        if (!createInfo.CasterGUID.IsGameObject())
```
Comment (Shauren) [inline src/server/game/Spells/Auras/SpellAuras.cpp:397]:
```suggestion
        if (createInfo.CasterGUID.IsUnit())
```

File: src/server/game/Spells/Auras/SpellAuras.cpp [hunk 4]
```diff
@@ -527,6 +522,11 @@ Unit* Aura::GetCaster() const
     return ObjectAccessor::GetUnit(*GetOwner(), GetCasterGUID());
 }
 
+WorldObject* Aura::GetWorldObjectCaster() const
+{
+    return ObjectAccessor::GetWorldObject(*GetOwner(), GetCasterGUID());
```
Comment (Shauren) [inline src/server/game/Spells/Auras/SpellAuras.cpp:527]:
```suggestion
    if (GetCasterGUID().IsUnit())
        return GetCaster();
    return ObjectAccessor::GetWorldObject(*GetOwner(), GetCasterGUID());
```

File: src/server/game/Spells/Auras/SpellAuras.cpp [hunk 5]
```diff
@@ -527,6 +522,13 @@ Unit* Aura::GetCaster() const
     return ObjectAccessor::GetUnit(*GetOwner(), GetCasterGUID());
 }
 
+WorldObject* Aura::GetWorldObjectCaster() const
+{
+    if (GetCasterGUID().IsUnit())
+        return GetCaster();
+    return ObjectAccessor::GetWorldObject(*GetOwner(), GetCasterGUID());
```
Comment (Shauren) [inline src/server/game/Spells/Auras/SpellAuras.cpp:529]:
```suggestion
        return GetCaster();

    return ObjectAccessor::GetWorldObject(*GetOwner(), GetCasterGUID());
```

**Pull request merged**